pineappl_capi/src/lib.rs

```diff
@@ -1867,3 +1895,77 @@ pub unsafe extern "C" fn pineappl_grid_convolve(
         mu_scales,
     ));
 }
+
+/// Get the number of convolutions for a given Grid.
+///
+/// # Safety
+///
+/// If `grid` does not point to a valid `Grid` object, for example when `grid` is the null pointer,
+/// this function is not safe to call.
+#[no_mangle]
+pub unsafe extern "C" fn pineappl_convolutions_len(grid: *mut Grid) -> usize {
+    let grid = unsafe { &mut *grid };
+    grid.convolutions().len()
+}
+
+/// Get the shape of a subgrid for a given bin, channel, and order.
+///
+/// # Safety
+///
+/// If `grid` does not point to a valid `Grid` object, for example when `grid` is the null pointer,
+/// this function is not safe to call. Additionally, the pointer that specifies the shape of the
+/// subgrid must be an array whose size must be `(n+1)` where `n` represents the number of
+/// convolutions as given by `pineappl_convolutions_len`.
+#[no_mangle]
+pub unsafe extern "C" fn pineappl_subgrid_shape(
+    grid: *mut Grid,
+    bin: usize,
+    order: usize,
+    channel: usize,
+    shape: *mut usize,
+) {
+    let grid = unsafe { &mut *grid };
+    let mut subgrid = grid.subgrids()[[order, bin, channel]].clone();
+
+    let subgrid_shape = if subgrid.is_empty() {
+        let subgrid_dim = grid.convolutions().len() + 1;
+        &vec![0; subgrid_dim]
+    } else {
+        subgrid.shape()
+    };
+
+    let shape = unsafe { slice::from_raw_parts_mut(shape, grid.convolutions().len() + 1) };
```

Comment:
That should be `grid.kinematics().len()` instead of `grid.convolutions().len() + 1`, because we may (and do) have grids with more than one scale.

Given that, we probably need to provide `pineappl_grid_kinematics_len`.

Reply:
Good point! You're completely right, that was an oversight. Fixed now in [bd68162](https://github.com/NNPDF/pineappl/pull/337/commits/bd68162dd7c88b9961ccbbf9641804cf2a0a2764)!